File: problems/13_1d_dp/0139_word_break/solution.dart
// 0139. Word Break
bool wordBreak(String s, List<String> wordDict) {
  final words = wordDict.toSet();
  var maxW = 0;
  for (final w in words) if (w.length > maxW) maxW = w.length;
  final n = s.length;
  final dp = List<bool>.filled(n + 1, false);
  dp[0] = true;
  for (var i = 1; i <= n; i++) {
    final lo = i - maxW < 0 ? 0 : i - maxW;
    for (var j = lo; j < i; j++) {
      if (dp[j] && words.contains(s.substring(j, i))) { dp[i] = true; break; }
    }
  }
  return dp[n];
}
void main() {
  if (wordBreak("leetcode", ["leet", "code"]) != true) throw 'a';
  if (wordBreak("applepenapple", ["apple", "pen"]) != true) throw 'b';
  if (wordBreak("catsandog", ["cats", "dog", "sand", "and", "cat"]) != false) throw 'c';
  if (wordBreak("a", ["a"]) != true) throw 'd';
  if (wordBreak("ab", ["a"]) != false) throw 'e';
  if (wordBreak("aaaaaaa", ["aaaa", "aaa"]) != true) throw 'f';
  print('All tests passed!');
}

File: problems/13_1d_dp/0139_word_break/solution.py
"""
Problem: 0139. Word Break
"""
from typing import List


class Solution:
    def wordBreak(self, s: str, wordDict: List[str]) -> bool:
        words = set(wordDict)
        max_w = max((len(w) for w in words), default=0)
        n = len(s)
        dp = [False] * (n + 1)
        dp[0] = True
        for i in range(1, n + 1):
            for j in range(max(0, i - max_w), i):
                if dp[j] and s[j:i] in words:
                    dp[i] = True
                    break
        return dp[n]


if __name__ == "__main__":
    sol = Solution()
    assert sol.wordBreak("leetcode", ["leet", "code"]) is True
    assert sol.wordBreak("applepenapple", ["apple", "pen"]) is True
    assert sol.wordBreak("catsandog", ["cats", "dog", "sand", "and", "cat"]) is False
    assert sol.wordBreak("a", ["a"]) is True
    assert sol.wordBreak("ab", ["a"]) is False
    assert sol.wordBreak("aaaaaaa", ["aaaa", "aaa"]) is True
    print("All tests passed!")
